File: medium/2615-sum-of-distances/solution.java
import java.util.*;

/**
 * 2615. Sum of Distances
 * Time: O(n)
 * Space: O(n)
 */
class Solution {
    public long[] distance(int[] nums) {
        int n = nums.length;
        long[] arr = new long[n];

        // Group indices by value
        Map<Integer, List<Integer>> groups = new HashMap<>();
        for (int i = 0; i < n; i++) {
            groups.computeIfAbsent(nums[i], k -> new ArrayList<>()).add(i);
        }

        // For each group, calculate distances using prefix sum
        for (List<Integer> indices : groups.values()) {
            int count = indices.size();

            // Build prefix sum of indices
            long[] prefix = new long[count + 1];
            for (int j = 0; j < count; j++) {
                prefix[j + 1] = prefix[j] + indices.get(j);
            }

            for (int j = 0; j < count; j++) {
                long idx = indices.get(j);
                // Left contribution
                long left = idx * j - prefix[j];
                // Right contribution
                long right = (prefix[count] - prefix[j + 1]) - idx * (count - j - 1);
                arr[(int) idx] = left + right;
            }
        }

        return arr;
    }
}

File: medium/2615-sum-of-distances/solution.py
from typing import List
from collections import defaultdict


class Solution:
    def distance(self, nums: List[int]) -> List[int]:
        """
        2615. Sum of Distances
        Time: O(n)
        Space: O(n)
        """
        n = len(nums)
        arr = [0] * n

        # Group indices by value
        groups = defaultdict(list)
        for i, num in enumerate(nums):
            groups[num].append(i)

        # For each group, calculate distances using prefix sum
        for indices in groups.values():
            # Build prefix sum of indices
            prefix = [0]
            for idx in indices:
                prefix.append(prefix[-1] + idx)

            count = len(indices)
            for j, idx in enumerate(indices):
                # Left contribution: j * idx - sum of indices to the left
                left = idx * j - prefix[j]
                # Right contribution: sum of indices to the right - (count - j - 1) * idx
                right = (prefix[count] - prefix[j + 1]) - idx * (count - j - 1)
                arr[idx] = left + right

        return arr
